### DISCUSSION TITLE
Workspace Setup

### DISCUSSION THREAD

**Discussion by aarmijo** [denkbares/d3web-KnowWE#7] [Q&A] [answered]
Hi there,

I'm trying to set-up the project and there seems to be a problem in the deployment into tomcat. Could you, please, provide some hints on how to solve the dependencies.

I get this log:

feb 08, 2022 5:55:39 P. M. [JSPWiki] de.knowwe.core.Environment init
GRAVE: Invalid initialization of the wiki. This is caused by an invalid wiki plugin. Wiki is in unstable state. Please exit and correct before using the wiki.
java.lang.Error: Unresolved compilation problems: 
	The package javax.xml is accessible from more than one module: <unnamed>, java.xml
	The package javax.xml.transform is accessible from more than one module: <unnamed>, java.xml
	The package javax.xml.transform.stream is accessible from more than one module: <unnamed>, java.xml
	The package javax.xml.validation is accessible from more than one module: <unnamed>, java.xml
	The package javax.xml.validation is accessible from more than one module: <unnamed>, java.xml
	The package javax.xml.validation is accessible from more than one module: <unnamed>, java.xml
	The package org.xml.sax is accessible from more than one module: <unnamed>, java.xml
	The package org.xml.sax is accessible from more than one module: <unnamed>, java.xml
	Source cannot be resolved to a type
	The method loadAsSource(WikiAttachment) from the type XMLValidationHandler refers to the missing type Source
	Schema cannot be resolved to a type
	SchemaFactory cannot be resolved to a type
	SchemaFactory cannot be resolved
	SAXException cannot be resolved to a type
	Source cannot be resolved to a type

Thanks in advance.
Alberto

**Comment by AlbrechtStriffler** [comment, marked as answer]
Hi Alberto,
at first glance, I would guess that you are using an incompatible Tomcat or Java version in your setup. We are usually running Tomcat 9.0.44 and Java 14, so if you have different versions, try these next. If that doesn't solve the problem, I need a bit more information, e.g. more of your log.

Regards
Albrecht

**Reply by aarmijo** [reply]
Hi Albrecht,
I've been testing out some things and this is what I found out:
1 - The project builds ok using external maven and installed jdk 17
2 - It builds ok in intellij idea ultimate and the debugging capabilities using a tomcat 9 server
3 - In eclipse 2021 (latest at the time of writting) it builds ok but the deployment in tomcat does not work out of the box

I followed some guidelines from here:
https://www.d3web.de/Wiki.jsp?page=How-To%20Create%20Your%20Own%20Workspace

I guess you currently use intellij idea for development but if you also used eclipse it would be good if you could update the guidelines at any point.

Regards and thank you,
Alberto

**Reply by AlbrechtStriffler** [reply]
Hi Alberto,

yes, in our team, we are all using IntelliJ Idea at the moment, so the eclipse part of the guide is very much out of date unfortunately (2014 I think). Did you manage to get in running in eclipse? If so, feel free to drop some hints what other people would have to do, I will try and update the guide accordingly.

Regarding Java 17: Yes, the core of KnowWE (open-source part) should run fine using Java 17. Since many of use are running other modules and additional closed source plugins not yet compatible with Java 17 in the same IntelliJ project, we are currently trying to keep everything compatible with the Java 14 level for now.

Best regards and thanks for the feedback!
Albrecht

**Reply by aarmijo** [reply]
Hi Albrecht,

Yes, I've got it up and running in Eclipse in Windows. This is my set-up:

Adoptium Temurin 17 JDK
https://github.com/adoptium/temurin17-binaries/releases/download/jdk-17.0.2%2B8/OpenJDK17U-jdk_x64_windows_hotspot_17.0.2_8.msi

Tomcat 9.0.26
https://tomcat.apache.org/download-90.cgi

Eclipse 2022-03
https://www.eclipse.org/downloads/packages/release/2022-03/m2

Maven 3.6.1. The embedded maven in eclipse seems to have some issues when constraining the compilation to JDK14
https://archive.apache.org/dist/maven/maven-3/

Instructions:
1.	Import d3web-KnowWE project as a Maven project. File -> Import -> Existing Maven Projects
2.	Window -> Preferences -> Java -> Installed JREs. Add Adoptium JDK and select it
3.	Make sure that the ‘JRE System Library’ of each maven project is set to ‘JavaSE-14’ compliance level. Also make sure that in maven project properties, the JDK compliance is limited to 14 in the ‘Java Compiler’ section
4.	In KnowWE-App project, under Maven Dependencies properties -> Use Maven Project settings -> Active Maven Profiles, set ‘core-plugins’ as active profile. This will copy the core dependencies to the KnowWE-App application when deploying it into Tomcat
5.	Under KnowWE-App project properties -> Java Build Path -> Source, add the next Folder: src/resources/local/WEB-INF/classes. This will copy the jspwiki.properties to the KnowWE-App application when deploying it into Tomcat
6.	Add the run configuration to build the project: Run -> Run Configurations -> Double click on Maven Build to create a new configuration. Set the Base directory: ${workspace_loc:/KnowWE}; Goals: install; Profiles: core-plugins; Skip Tests. Under JRE -> Execution environment, select JavaSE-14. Select Maven Runtime -> External (It seems that EMBEDDED flavour is not well managed by Eclipse and the compilation is not constrained to version 14)
7.	Click Run to build the project
8.	Under the tab Servers, add a Tomcat9 web application server and select the location where you downloaded it. Add KnowWE-App as resource to the right ‘Configured’ section. Under Edit Server Runtime select as JRE the installed JDK 1.7.
9.	Click Start to launch the server. Open the application in http://localhost:8080/KnowWE-App

Hope it helps.

Regards and thank you,
Alberto

**Reply by AlbrechtStriffler** [reply]
Hi Alberto,
thanks a lot for your setup-guide! I added it on https://www.d3web.de/Wiki.jsp?page=How-To%20Create%20Your%20Own%20Workspace.
I also updated the IntelliJ part a bit, but it isn't quite done yet. If I have a bit more time, I will attempt to create an additional screencast...
Regards
Albrecht
